control-base/src/main/java/fi/nls/oskari/control/feature/GetWFSFeaturesHandler.java

```diff
@@ -0,0 +1,236 @@
+    public void init() {
+        if (permissionHelper != null) {
+            permissionHelper = new PermissionHelper(
+                    ServiceFactory.getMapLayerService(),
+                    ServiceFactory.getPermissionsService());
+        };
+        if (wfsClient != null) {
+            wfsClient = new OskariWFS110Client();
+        }
+    }
+
+    @Override
+    public void handleAction(ActionParameters params) throws ActionException {
+        int layerId = params.getRequiredParamInt(ActionConstants.PARAM_ID);
+        Envelope bbox = parseBbox(params.getRequiredParam(PARAM_BBOX));
+        OskariLayer layer = findLayer(layerId, params.getUser());
+        // Only support WebMercator for now
+        // String targetSrs = params.getRequiredParam(ActionConstants.PARAM_SRS);
+        String targetSrsName = "EPSG:3857";
+
+        // TODO: Figure out if layer supports targetSrsName
+        // If it does let the WFS service do the transformation
+        SimpleFeatureCollection fc = getFeatures(layer, bbox, requestSrsName, targetSrsName);
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            if (fc.isEmpty()) {
+                baos.write(EMPTY_GEOJSON_FEATURE_COLLECTION);
+            } else {
+                FJ.writeFeatureCollection(fc, baos);
+            }
+            ResponseHelper.writeResponse(params, 200, GEOJSON_CONTENT_TYPE, baos);
+        } catch (IOException e) {
+            ResponseHelper.writeError(params, "Failed to write GeoJSON");
+        }
+    }
+
+    private Envelope parseBbox(String bbox) throws ActionParamsException {
+        String[] a = bbox.split(",", 4);
+        if (a.length != 4) {
+            throw new ActionParamsException(ERR_BBOX_INVALID);
+        }
+        try {
+            return new Envelope(
+                    Double.parseDouble(a[0]),  // x1
+                    Double.parseDouble(a[2]),  // x2
+                    Double.parseDouble(a[1]),  // y1
+                    Double.parseDouble(a[3])); // y2
+        } catch (NumberFormatException e) {
+            throw new ActionParamsException(ERR_BBOX_INVALID);
+        }
+    }
+
+    private OskariLayer findLayer(int layerId, User user) throws ActionException {
+        OskariLayer layer = permissionHelper.getLayer(layerId, user);
+        if (!OskariLayer.TYPE_WFS.equals(layer.getType())) {
+            throw new ActionParamsException(ERR_LAYER_TYPE_NOT_WFS);
+        }
+        return layer;
+    }
+
+    protected SimpleFeatureCollection getFeatures(OskariLayer layer, Envelope bbox,
+            String requestSrsName, String targetSrsName) throws ActionException {
+        CoordinateReferenceSystem requestCRS;
+        CoordinateReferenceSystem targetCRS;
+        try {
+            requestCRS = CRS.decode(requestSrsName, true);
+            targetCRS = CRS.decode(targetSrsName, true);
+        } catch (Exception e) {
+            throw new ActionException(ERR_CRS_DECODE_FAIL);
```

Comment:
This should be ActionParamsException since we don't pass the cause and either way that might not be too interesting.

Reply:
^Just to prevent excessive logging.